### ISSUE TITLE
Support "strict mode" for RFC7159 parsing

### ISSUE THREAD

**Issue by jskeet** [JamesNK/Newtonsoft.Json#646]
Currently Json.NET parses in a more lenient manner than RFC7159. For example:

``` json
{
  "foo": Infinity
}
```

... will be parsed by Json.NET, but isn't valid under RFC7159.

I'm sure that's generally useful, but it makes it harder to build on Json.NET when strict parsing is required. Other features such as date detection can be disabled already, but I don't _think_ there's currently any way of disabling this leniency.

This is only one example - there may well be others that James is aware of, given that he's rather more deeply "in" this world than I am :) (Comments, possibly the quote characters used, etc.)

**Comment by JamesNK** [comment]
What you could do is inherit from JsonTextReader, override Read and check the incoming data after each read. If QuoteChar is ' then throw an error, if Value is a double and is double.Infinity/double.NaN then throw an error, etc.

**Comment by jskeet** [comment]
Thanks, that's definitely a practical workaround. I think this would be generally useful functionality to be in Json.NET itself at some point, but I'll give that a try for now. (I'll try to identify the ways in which Json.NET is more lenient than the RFC, and note what I find here for future readers...)

**Comment by CyrusNajmabadi** [comment]
> (I'll try to identify the ways in which Json.NET is more lenient than the RFC, and note what I find here for future readers...

@jskeet I'm adding JSON editor support for Roslyn over here: https://github.com/dotnet/roslyn/pull/24110

I've been using json.net to base how parsing will work (as that's likely the json parser that most of them wil be using in practice).  However, i think it would be nice to allow a strict mode as well.  There are very obvious ways that json.net deviates from standard JSON.  But i'm concerned i may miss other cases.  Do you happen to have the list that identifies all the ways Json.Net is more lenient?  That would be very helpful for ensuring i don't miss anything.

**Comment by JamesNK** [comment]
Off the top of my head:

- Properties with single quotes instead of double quotes
- Properties without quotes
- NaN, Infinity, -Infinity
- JavaScript constructors, e.g. new DateTime(12233434)
- Single and multi-line comments
- Trailing commas
- Octal numbers (number starts with a zero)
- Hexadecimal numbers (I think...)
- New lines in strings

**Comment by JamesNK** [comment]
https://github.com/JamesNK/Newtonsoft.Json/tree/master/Src/Newtonsoft.Json.Tests/JsonTextReaderTests

**Comment by CyrusNajmabadi** [comment]
@JamesNK Thanks for the list.  Fortunately, reading the actual json spec was simple enough that i was able to find all that as well.  Other things i can think of are leniency around numbers and whitespace.

**Comment by CyrusNajmabadi** [comment]
Interestingly enough, i haven't found an actual .net json parser that tries to match the spec precisely.  I've found deviations in several (including MS's own DataContractJsonSerializer and JavaScriptSerializer).  

Note: i'm a big fan of json.net and how it's addressed this space (including accepting a superset of json).  Do you think there's a place in json.net to have an option for more strict processing of json?  If so, i'd be happy to contribute.  But i don't know if such an option fits your vision for the project.

Thanks!

**Comment by CyrusNajmabadi** [comment]
> Hexadecimal numbers (I think...)

Yup.  Not allowed in json.  Octal isn't allowed either.  Leading zeros are also verboten.  Fun stuff.

**Comment by JamesNK** [comment]
> Do you think there's a place in json.net to have an option for more strict processing of json?

I'm not sure. I get the feeling that if I add a strict mode as just a bool flag on JsonTextReader then people will start asking for more options. "I want to error on comments but still accept trailing commas", etc. I don't want to add lots of fields to JsonTextReader to configure how strict it should be.

**Comment by PR4GM4** [comment]
+1 This is needed. The reason case on my end is that I use a "Repository" for json files that the public can upload and download. PHP uses the proper RFC implementation so if the user makes a JSON that works fine with the C# tester application, they think there wont be any issues, but in fact, there will be when its trying to be parsed and used on my PHP upload script.

So a strict mode here, would be insanely useful to tell the user on my C# end that its not proper syntax, and they can google or lint check to figure out there issue. Otherwise they just say my PHP script is broken plz fix m8 bla bla annoying shit when in fact, its there JSON thats broken.

**Comment by JonHanna** [comment]
>  I get the feeling that if I add a strict mode as just a bool flag on JsonTextReader then people will start asking for more options.

I think there's a different dimension of change between "valid by the standard" and "valid by the standard except for in the following cases…" that makes this request different to a request like that.

What I could see happening though is a request for validity according to some RFC 8259bis that replaces RFC 8259 (indeed, the relevant RFC was 7159 when this issue was opened, though I think the only change in validity was that UTF-8 is required for interchange which is at a different level than parsing anyway) and/or a later version of ECMA 404 / ISO/IEC 21778 (presumably the three standards will continue to track each other, but the timings of versions could differ). It might be wise future-proofing to define validity through an enum that could have later not-fully-compatible standard versions added to it, but not to offer such validation for anything other than a published standard.

But a validating reader could also be built as a separate package.

**Comment by PR4GM4** [comment]
Just add a bool flag, honestly if people ask for more options, we can discuss if it will be required or not.

**Comment by TylerBrinkley** [comment]
You could add a flag enum to represent the allowable deviations. If you make it int based it would support up to 32 different options but 64 different options if you make it long based.

**Comment by CyrusNajmabadi** [comment]
Personally, i'd have no problem just adding a strict mode without any flexibility.  If you want any form of looseness, you don't ask for strict.  If you want strict, you get it.  But i'm not the project owner :)

**Comment by sixlettervariables** [comment]
@cyrusnajmabadi Manatee.Json only supports strict mode (I believe).

**Comment by elgonzo** [comment]
> I'm not sure. I get the feeling that if I add a strict mode as just a bool flag on JsonTextReader then people will start asking for more options. "I want to error on comments but still accept trailing commas", etc. I don't want to add lots of fields to JsonTextReader to configure how strict it should be.

You don't need to add lots of fields just because someone is asking. ;)

Introducing a (bool) option for enabling strict mode would help many users. It makes sense, as it allows users to read/verify standard-conform Json data without the hassle of building and maintaining unwieldy JsonTextReader.Read() overrides.

For anyone else wanting an individual blend of a little strict here and a little loose there, just refer them to your comment from Sept. 2015:
> What you could do is inherit from JsonTextReader, override Read and check the incoming data after each base.Read().

**Comment by guylando** [comment]
What about deserializing and then serializing and comparing to original string? This validates that properties are surrounded by ".." for example.

**Comment by jskeet** [comment]
@guylando: That would be a very inefficient way of validating, IMO.

**Comment by asthomas** [comment]
For what it's worth, I am running into exactly this issue.  I need to determine whether a document is valid JSON.  A strict mode would be perfect for that.

The whole purpose of a spec is to ensure interoperability.  If different implementations deviate from the spec, they are no longer interoperable.  In my case interoperability is paramount.

**Comment by gdalsnes** [comment]
I think DateParseHandling.None should also be a part of strict mode. Strict mode would make the road from Json.Net to System.Text.Json less bumpy.
